add test for binding issues

jbblanchet opened the pull request:
This was a bit more complex than anticipated, because I had to add babel to reproduce the issue, but this test will fail if the regression fixed by #125 comes back.

This also brings to the forefront the difference between components created by inheritance, and components created by createClass. ~~The re-binding will give an error on the latter. Not sure what a long term solution is.~~ This can be fixed by looking at the prototype of the element.

jbblanchet commented:
This one should be ready to merge, but I'd still like to add tests to check all the possibilities (createClass vs. React.Component vs. instance method vs. props etc.). I'll either update this PR if not merged yet or create a new one when I have time to add those later today.

Pomax commented on package.json:
```diff
@@ -30,9 +30,11 @@
     "object-assign": "^4.0.1"
   },
   "devDependencies": {
+    "babel-preset-es2015": "^6.14.0",
```
why do we need these? `isPrototypeOf` has been in JavaScript since ECMAScript 3rd edition, so no babelification should be necessary here?

Pomax commented on test/.eslintrc.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "parserOptions": {
+    "ecmaVersion": 6
```
no ES6 is used in the code, so this should not be necessary

jbblanchet replied:
@Pomax we're using classes and inheritance to be able to reproduce the issue.

jbblanchet replied:
@Pomax But only in the tests.

Pomax replied:
left a note there, too. I don't see why the ES6 rewrite was necessary. Stateless components are simply functions with signature `function(props) { return whateveryouneedrendered }`.

Pomax commented on test/test.js:
```diff
@@ -69,33 +69,30 @@ describe('onclickoutside hoc', function() {
-    var Component = React.createClass({
-      getInitialState: function() {
-        return {
+    class Component extends React.Component {
```
why not use a proper stateless component, which is just a `function(props) { return UICodeGoesHere }`?

jbblanchet replied:
@Pomax A stateless component might be sufficient to generate the error, but I do think that testing the 3 ways to create a component (stateful component with inheritance, createClass and stateless function) will give you the best chance to catch these errors early next time.

I should be able to add those tests by the end of the day. If you prefer that I only add tests for stateless vs createClass, let me know.

jbblanchet replied:
Still thinking about your comment, and stateless functions won't be re-bound here, because they do not inherit from React.Component. I'm not sure if it makes sense to support instance method on stateless functions, but if so, the test should be reversed to check if createClass was used. The (very ugly) way to do that appears to be:

```javascript
var ReactClassComponentPrototype = (Object.getPrototypeOf(Object.getPrototypeOf(new (React.createClass({ render: function () {} }))())))
```

Source: https://discuss.reactjs.org/t/how-to-determine-if-js-object-is-react-component/2825/5

jbblanchet replied:
But (still thinking out loud) I don't think it makes sense to auto bind on the instance on stateless functions. The whole point of those is to be... stateless. So the logic here would be:
- If it has been created by inheritance: re-bind.
- If it has been created with createClass: binding already taken care of, do nothing.
- If it it a stateless method: binding shouldn't be necessary, do nothing.

But to test that logic, I do need to use es6 classes, unless I try and mock the inheritance, but that seems risky and more complex.

jbblanchet commented:
@Pomax since it seems that there still should be a discussion on the tests, and that the actual fix should be merged as soon as possible (because 5.5 introduced a breaking change over 5.3), do you want me to create a separate PR for the fix so you can release it ASAP, then update this one with the tests?

jbblanchet commented:
Added some more tests to cover the use case I could think of. Components are now created by inheritance (with instance method and prop), with a createClass function (with instance method and prop) and with a stateless function (with prop). That last test is not working, because stateless functions do not support refs, so I skipped it for now. It also made me realize that re-binding was not necessary on props, since props should not be bound to the instance, so I updated the code.

Pomax commented:
new code yields a pretty readable `npm test` log so I'm happy to merge this in.

Pomax commented:
v5.6.0 released
